splitio/javascript-sync-tools
# Set Storage from CLI & added other CLI options

# Javascript Slim Synchronizer

## What did you accomplish?
In this PR, we are adding:
1. CLI option to define and set the Storage.
2. Added CLI capabilities to set APIKEY, APIURL and EVENT_API_URL through CLI param.
3. Added CLI capabilities to set `ImpressionsMode`
4. Added CLI capabilities to enable Logger debug Mode
5. Added ImpressionsCount Synchroniser

I've also improved/updated some tests and the information that is logged through the execution. Removed some `todos`.

## How do we test the changes introduced in this PR?
There are a couple of stuff to check

1. Run the usual tests
`npm run tests`

2. Test (manually) the build 
You may want to change the different ways to test this app.
First, build the app: `npm run build`

Then, you can manually test the following scenarios:

### Using CLI parameters: 
`./dist/cli.js -k YOURAPIKEY -s ../src/__tests__/customStorage/InMemoryStorage.js -r https://sdk.split-stage.io/api -e https://events.split-stage.io/api/`

#### Enabling Logger in Debug mode:
`./dist/cli.js -k YOURAPIKEY -s ../src/__tests__/customStorage/InMemoryStorage.js -r https://sdk.split-stage.io/api -e https://events.split-stage.io/api/ -d`

You should see some Logger's log in the console.

#### ImpressionsMode DEBUG:
`./dist/cli.js -k YOURAPIKEY -s ../src/__tests__/customStorage/InMemoryStorage.js -r https://sdk.split-stage.io/api -e https://events.split-stage.io/api/ -i`

If DEBUG mode is set, then the ImpressionsCount Synchroniser won't be executed.

### Using Environment variables (`env` mode): 
`./dist/cli.js -m env -s ../src/__tests__/customStorage/InMemoryStorage.js`

You can create a local `.env` file
```
APIKEY=YOURAPIKEY
API_URL=https://sdk.split-stage.io/api
EVENTS_API_URL=https://events.split-stage.io/api/
```

### Using JSON file (`json` mode)
note: Create a JSON file
```
{
	"APIKEY": "YOURAPIKEY",
	"API_URL": "https://sdk.split-stage.io/api"
	"EVENTS_API_URL": "https://events.split-stage.io/api/"
}
```

`./dist/cli.js -m json --json-file ../pathTo/configs.json -s ../src/__tests__/customStorage/InMemoryStorage.js `


## Related JIRA tickets and PRs
[[Synchroniser] - Ready Hook for Storage](https://splitio.atlassian.net/browse/SDKS-2632)
[[Pluggable storages] JavaScript lightweight synchronizer](https://splitio.atlassian.net/browse/SDKS-2620)

## Extra Notes
